apps/builder/app/builder/features/command-panel/token-group.tsx

```diff
@@ -0,0 +1,224 @@
+    return instanceSelector;
+  }
+  const instance = instances.get(instanceId);
+  if (instance) {
+    for (const child of instance.children) {
+      if (child.type === "id") {
+        const matched = findInstanceById(
+          instances,
+          [child.value, ...instanceSelector],
+          targetId
+        );
+        if (matched) {
+          return matched;
+        }
+      }
+    }
+  }
+};
+
+const selectToken = (
+  instanceId: Instance["id"],
+  tokenId: StyleSource["id"]
+) => {
+  const instances = $instances.get();
+  const pagesData = $pages.get();
+  if (pagesData === undefined) {
+    return;
+  }
+  const pages = [pagesData.homePage, ...pagesData.pages];
+  for (const page of pages) {
+    const instanceSelector = findInstanceById(
+      instances,
+      [page.rootInstanceId],
+      instanceId
+    );
+    if (instanceSelector) {
+      $awareness.set({ pageId: page.id, instanceSelector });
+      const selectedStyleSources = new Map($selectedStyleSources.get());
+      selectedStyleSources.set(instanceId, tokenId);
+      $selectedStyleSources.set(selectedStyleSources);
+      break;
+    }
+  }
+};
+
+type InstanceOption = {
+  label: string;
+  id: string;
+};
+
+const TokenInstances = ({ tokenId }: { tokenId: StyleSource["id"] }) => {
+  const usages = useStore($styleSourceUsages);
+  const usedInInstanceIds = usages.get(tokenId) ?? new Set();
+  const instances = $instances.get();
+  const metas = $registeredComponentMetas.get();
+  const usedInInstances: InstanceOption[] = [];
+  for (const instanceId of usedInInstanceIds) {
+    const instance = instances.get(instanceId);
+    const meta = metas.get(instance?.component ?? "");
+    if (instance && meta) {
+      usedInInstances.push({
+        label: getInstanceLabel(instance, meta),
+        id: instance.id,
+      });
+    }
+  }
+  const [search, setSearch] = useState("");
+  let matches = usedInInstances;
+  // prevent searching when value is empty
+  // to preserve original items order
+  if (search.trim().length > 0) {
+    for (const word of search.trim().split(/\s+/)) {
+      matches = matchSorter(matches, word, {
+        keys: ["label"],
+      });
+    }
+  }
+  return (
+    <>
+      <CommandInput value={search} onValueChange={setSearch} />
+      <Flex direction="column" css={{ maxHeight: 300 }}>
+        <ScrollArea>
+          <CommandList>
+            {matches.map(({ id, label }) => (
+              <CommandItem
+                key={id}
+                // preserve selected state when rerender
+                value={id}
+                onSelect={() => {
+                  selectToken(id, tokenId);
+                  closeCommandPanel();
+                }}
+              >
+                <Text variant="labelsTitleCase">{label}</Text>
+              </CommandItem>
+            ))}
+          </CommandList>
+        </ScrollArea>
+      </Flex>
+    </>
+  );
+};
+
+export const TokenGroup = ({ options }: { options: TokenOption[] }) => {
```

Comment:
Why do we need all these token specific components? Why can't tokens, instances, or variables search all be using the same components?

Reply:
Caz all of them have different logic to implement, different groups. "Different functions or functions with conditions inside."

Reply:
you mean for actions? that could be a config per group that has an action name and action handler

Reply:
Each group look differently. Has different requirements. We don't need to make it look the same.

Reply:
Hmm It sounds like lots of code while we don't know if we need this level of customization, I would expect different categories to have different actions, but nothing else

Reply:
<img width="166" alt="image" src="https://github.com/user-attachments/assets/e69b8e3b-7a7a-4da0-87b3-a41ec634ab9a" />
<img width="426" alt="image" src="https://github.com/user-attachments/assets/b51ac416-c44a-4329-be06-dc5e47535894" />
<img width="127" alt="image" src="https://github.com/user-attachments/assets/30dd2beb-5526-49bd-ba10-c7d1656e5a67" />